File: apps/app/lib/features/orders/domain/entities/order.dart
import 'package:equatable/equatable.dart';

import '../../../address/domain/entities/address.dart';

/// One line item on an order (embedded array on the order document).
class OrderItem extends Equatable {
  final String productId;
  final String productName;
  final String productImage;
  final String variantId;
  final String variantName;
  final String variantColor;
  final double mrp;
  final double offerPrice;
  final int quantity;
  final double subtotal;

  const OrderItem({
    required this.productId,
    required this.productName,
    required this.productImage,
    required this.variantId,
    required this.variantName,
    required this.variantColor,
    required this.mrp,
    required this.offerPrice,
    required this.quantity,
    required this.subtotal,
  });

  @override
  List<Object?> get props => [
        productId,
        productName,
        productImage,
        variantId,
        variantName,
        variantColor,
        mrp,
        offerPrice,
        quantity,
        subtotal,
      ];
}

/// One entry in an order's status history (embedded array).
class StatusTimeline extends Equatable {
  final String status;
  final DateTime? timestamp;
  final String note;

  const StatusTimeline({
    required this.status,
    required this.timestamp,
    required this.note,
  });

  @override
  List<Object?> get props => [status, timestamp, note];
}

/// A customer order document (`orders/{id}`).
///
/// Money fields are recomputed server-side and never trusted from the client;
/// the app reads them for display only.
class Order extends Equatable {
  final String id;
  final String userId;
  final String userName;
  final String userPhone;
  final String userEmail;
  final List<OrderItem> items;
  final double subtotal;
  final double couponDiscount;
  final String couponCode;
  final double walletAmountUsed;
  final double deliveryCharge;
  final double gstAmount;
  final double grandTotal;

  /// 'Razorpay' | 'Wallet' | 'Both'.
  final String paymentMethod;

  /// 'Pending' | 'Paid' | 'Failed' | 'Refunded'.
  final String paymentStatus;
  final String razorpayOrderId;
  final String razorpayPaymentId;

  /// Order status enum ('Pending' … 'Delivered' | 'Cancelled').
  final String status;
  final Address shippingAddress;
  final String trackingId;
  final String courierName;

  /// Not in the shared schema — defaults to '' (see report).
  final String invoiceUrl;
  final List<StatusTimeline> statusTimeline;
  final String cancelReason;
  final bool isComboOrder;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const Order({
    required this.id,
    required this.userId,
    required this.userName,
    required this.userPhone,
    required this.userEmail,
    required this.items,
    required this.subtotal,
    required this.couponDiscount,
    required this.couponCode,
    required this.walletAmountUsed,
    required this.deliveryCharge,
    required this.gstAmount,
    required this.grandTotal,
    required this.paymentMethod,
    required this.paymentStatus,
    required this.razorpayOrderId,
    required this.razorpayPaymentId,
    required this.status,
    required this.shippingAddress,
    required this.trackingId,
    required this.courierName,
    required this.invoiceUrl,
    required this.statusTimeline,
    required this.cancelReason,
    required this.isComboOrder,
    required this.createdAt,
    required this.updatedAt,
  });

  int get totalItems => items.fold(0, (acc, item) => acc + item.quantity);

  @override
  List<Object?> get props => [
        id,
        userId,
        userName,
        userPhone,
        userEmail,
        items,
        subtotal,
        couponDiscount,
        couponCode,
        walletAmountUsed,
        deliveryCharge,
        gstAmount,
        grandTotal,
        paymentMethod,
        paymentStatus,
        razorpayOrderId,
        razorpayPaymentId,
        status,
        shippingAddress,
        trackingId,
        courierName,
        invoiceUrl,
        statusTimeline,
        cancelReason,
        isComboOrder,
        createdAt,
        updatedAt,
      ];
}

File: packages/shared/src/types/order.ts
import type { Timestamp } from 'firebase/firestore';
import type { OrderStatus } from '../config/enums';

/**
 * OrderItemProps / OrderProps — a customer order and its line items (spec §1.3, §1.7).
 *
 * WHY here (not per-app): admin, website, app and Cloud Functions all read/write the same order
 * documents; one definition keeps the contract identical. Money fields are recomputed server-side
 * (never trusted from the client). Date fields use `Timestamp` (not epoch millis).
 */
export interface OrderItemProps {
  productId: string;
  productName: string;
  productImage: string;
  variantId: string;
  variantName: string;
  variantColor: string;
  mrp: number;
  offerPrice: number;
  quantity: number;
  subtotal: number;
  /**
   * The GST rate applied to this line at checkout (18 = 18%), snapshotted from the variant so a later
   * rate change never rewrites a issued invoice. ABSENT on orders placed before per-line GST existed —
   * those carry only the order-level `gstAmount`, which is the whole-order figure at the then-global
   * rate. Treat absent as "no per-line breakdown available", not as 0%.
   */
  gstPercentage?: number;
  /** This line's share of `gstAmount`, in rupees. Absent alongside `gstPercentage`. */
  gstAmount?: number;
}

export interface OrderShippingAddress {
  fullName: string;
  phone: string;
  addressLine1: string;
  addressLine2: string;
  city: string;
  state: string;
  pincode: string;
  landmark: string;
}

export interface OrderStatusTimelineEntry {
  status: string;
  timestamp: Timestamp;
  note: string;
}

export interface OrderProps {
  id: string;
  userId: string;
  userName: string;
  userPhone: string;
  userEmail: string;
  items: OrderItemProps[];
  subtotal: number;
  couponDiscount: number;
  couponCode: string;
  walletAmountUsed: number;
  deliveryCharge: number;
  gstAmount: number;
  /**
   * True when `gstAmount` is CONTAINED IN the prices rather than added to them — so
   * `grandTotal = subtotal − couponDiscount + deliveryCharge`, and the invoice's taxable value is
   * `subtotal − gstAmount`.
   *
   * ABSENT on every order placed before 2026-08-04, when GST was charged on top
   * (`grandTotal = … + gstAmount`). Those invoices must keep stating what the customer actually paid,
   * so absent must be read as `false` — never defaulted to true. Use `isGstInclusive()`.
   */
  gstInclusive?: boolean;
  grandTotal: number;
  paymentMethod: 'Razorpay' | 'Wallet' | 'Both';
  paymentStatus: 'Pending' | 'Paid' | 'Failed' | 'Refunded';
  razorpayOrderId: string;
  razorpayPaymentId: string;
  status: OrderStatus;
  shippingAddress: OrderShippingAddress;
  trackingId: string;
  courierName: string;
  statusTimeline: OrderStatusTimelineEntry[];
  cancelReason: string;
  /**
   * Running total already refunded to the wallet by approved return requests against this order.
   * Server-written only (`approveReplacement`), and used as the cap so repeated partial returns can
   * never jointly refund more than `grandTotal`. Absent on orders with no approved return.
   */
  replacementRefundedTotal?: number;
  isComboOrder: boolean;
  keywords: string[];
  createdAt: Timestamp;
  updatedAt: Timestamp;
}
